doc/user-guide.md

```diff
@@ -1 +1,380 @@
+discuss them later.
+
+### Submodule Boundary Erasure
+
+Git submodules allow us to mount many external repositories into a
+singly-rooted tree, giving the approximation of a single mono-repo where
+selected subtrees are present (open).  Unfortunately, this abstraction is not
+well-serviced by most Git commands.  For example, say you have a meta-repo with
+a submodule named `x` that contains a file named `README.md`:
+
+```bash
+$ echo foo >> x/README.md
+```
+
+A naive attempt to stage this change results in error:
+
+```bash
+$ git add x/README.md
+fatal: Pathspec 'x/README.md' is in submodule 'x'
+```
+
+More generally, if you have modifications in many submodules, you can write:
+
+```bash
+$ git meta add .
+```
+
+staging all modified files in the mono-repo, treating the entire tree as one
+repository.  With plain Git, one would need to run `git add` in each submodule
+containing changes.  See the next section about the model git-meta uses for
+more information about what it means to stage a change to the mono-repo.
+
+The ability to address paths in the tree of the meta-repo as if the were a
+single repository (ignoring submodule boundaries) is provided by all git-meta
+commands that take paths, such as `reset`, `commit`, `checkout`, etc.
+
+### Better Submodule Interactions
+
+Historically, submodules provide a way for one repository to *reference* (not
+contain) another repository, and the entry for a submodule in a Git tree is a
+SHA indicating which commit is being referenced in the external repository.
+When you attempt to merge two commits containing different SHAs for the same
+submodule, Git simply flags the submodule as being conflicted.  This result
+makes sense based on the original use for submodules: on one branch we
+referenced one version of an external repository, on another branch we
+referenced a different version; we need a human to decide which one to use.
+While it might make sense to blindly accept a change to a commit that is a
+descendant of the original, it probably is not reasonable to e.g., create a
+merge commit -- this submodule is an external repository that we may not own.
+
+Unfortunately, this behavior renders Git commands that perform merges:
+`cherry-pick`, `merge`, `rebase`, etc. mostly useless when interacting with a
+mono-repo.  Instead of triggering a conflict when encountering submodule
+differences, we need these commands to:
+
+- `cherry-pick` -- perform a merge in the submodule
+- `merge` -- perform a merge in the submodule
+- `rebase` -- rebase submodule commits
+
+Furthermore, it may be necessary to fetch target commits during these
+operations.
+
+### Workarounds
+
+Workarounds are direct implementation of existing Git commands that provide no
+value other than to work, or to work faster.  One current example is
+`git submodule status`, that outputs the status of only a few
+submodules per second, making it unusable in repositories containing thousands
```

Comment:
I may have exaggerated a bit earlier, but I'm seeing 10/s with 2.11. Still means you're looking at 10 minutes for for 6k submodules